<a href="https://colab.research.google.com/github/DamianS18/Matma/blob/main/Kopia_notatnika_LA_Matrix_odwracanie_macierzy_ze_wzoru.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Odwracanie macierzy

Macierz odwrotna do macierzy $A$ to taka macierz $A^{-1}$, że $A \cdot A^{-1} = A^{-1} \cdot A = I$, gdzie $I$ to macierz jednostkowa.

**Przykład 1:**

In [ ]:
import sympy as sp #import sympy library
k1=sp.Matrix([[2, 1, 2], [5, 3, 1],[1, 1, 5]]) #stwórz macierz k1
k1

Matrix([
[2, 1, 2],
[5, 3, 1],
[1, 1, 5]])

In [ ]:
inverse_k1 = k1.inv() #oblicz odwrotność macierzy k1
inverse_k1

Matrix([
[7/4, -3/8, -5/8],
[ -3,    1,    1],
[1/4, -1/8,  1/8]])

In [ ]:
k1*inverse_k1 # sprawdzamy czy mnożenie macierzy przez jej odwrotność daje macierz jednostkową

Matrix([
[1, 0, 0],
[0, 1, 0],
[0, 0, 1]])

**Przykład 2:**

In [ ]:
import sympy as sp
a,b,c,d,e,f,g,h,i = sp.symbols('a b c d e f g h i') #definiujemy zmienne a-i
k=sp.Matrix([[a, b, c], [d, e, f], [g, h, i]]) #tworzymy macierz k
inverse_k = k.inv()
inverse_k

Matrix([
[ (e*i - f*h)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g), (-b*i + c*h)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g),  (b*f - c*e)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g)],
[(-d*i + f*g)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g),  (a*i - c*g)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g), (-a*f + c*d)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g)],
[ (d*h - e*g)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g), (-a*h + b*g)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g),  (a*e - b*d)/(a*e*i - a*f*h - b*d*i + b*f*g + c*d*h - c*e*g)]])

Ta skomplikowana postać wynika ze wzoru, który pozwala na obliczenie elementów macierzy odwrotnej. Zanim do niego przejdziemy, zdefiniujmy kilka pojęć.

**Transpozycja macierzy** jest operacją, która zamienia wiersze na kolumny. Transponowana macierz do macierzy $A$ to taka macierz $A^T$, że $A^T_{ij} = A_{ji}$.

In [ ]:
k

Matrix([
[a, b, c],
[d, e, f],
[g, h, i]])

In [ ]:
k.transpose() #transponujemy macierz k

Matrix([
[a, d, g],
[b, e, h],
[c, f, i]])

**Podwyznacznik** (minor) macierzy $A$ to wyznacznik macierzy, z której usunięto $i$-ty wiersz i $j$-tą kolumnę. Oznaczmy go $M_{ij}$.

In [ ]:
# Podwyznacznik (minor) macierzy $A$ to wyznacznik macierzy, z której usunięto $i$-ty wiersz i $j$-tą kolumnę. Oznaczmy go $M_{ij}$.
k2=sp.Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
k2

Matrix([
[1, 2, 3],
[4, 5, 6],
[7, 8, 9]])

In [ ]:
M01 = k2.minor_submatrix(0, 1) #obliczamy minor M01
M01

Matrix([
[4, 6],
[7, 9]])

In [ ]:
M22 = k2.minor_submatrix(2, 2)  #obliczamy minor M22
M22

Matrix([
[1, 2],
[4, 5]])

**Dopełnienie algebraiczne** elementu $a_{ij}= (-1)^{i+j} \, M_{ij}$, czyli podwyznacznik pomnożony przez $(-1)^{i+j}$.

**Rząd macierzy** to stopień największego, niezerowego minora tej macierzy. Rząd macierzy równy jest ilości liniowo niezależnych wierszy lub kolumn w macierzy.

In [ ]:
k2.rank()   #obliczamy rząd macierzy k2

2

### Macierz odwrotna 3x3

Wykorzystanie transpozycji macierzy dopełnień algebraicznych (minorów):

$$K^{-1} =
\begin{bmatrix}
a & b & c \\
d & e & f \\
g & h & i
\end{bmatrix}^{-1}
= \frac{1}{\det(\mathbf{K})}
\begin{bmatrix}
A & B & C \\
D & E & F \\
G & H & I
\end{bmatrix}^T
= \frac{1}{\det(\mathbf{A})}
\begin{bmatrix}
A & D & G \\
B & E & H \\
C & F & I
\end{bmatrix}
$$

Jeśli wyznacznik jest różny od zera, macierz jest odwracalna, a elementy macierzy pośredniej po prawej stronie powyższego wzoru są określone przez:

$$
\begin{aligned}
A &= det ( \begin{bmatrix} e & f \\ h & i \end{bmatrix} ) & B &= - det ( \begin{bmatrix} d & f \\ g & i \end{bmatrix} ) & C &= det ( \begin{bmatrix} d & e \\ g & h \end{bmatrix} ) \\
D &= - det ( \begin{bmatrix} b & c \\ h & i \end{bmatrix} ) & E &= det ( \begin{bmatrix} a & c \\ g & i \end{bmatrix} ) & F &= - det ( \begin{bmatrix} a & b \\ g & h \end{bmatrix} ) \\
G &= det ( \begin{bmatrix} b & c \\ e & f \end{bmatrix} ) & H &= - det ( \begin{bmatrix} a & c \\ d & f \end{bmatrix} ) & I &= det ( \begin{bmatrix} a & b \\ d & e \end{bmatrix} )
\end{aligned}
$$
czyli przez wyznaczniki macierzy 2x2, które powstają przez usunięcie odpowiednich wierszy i kolumn oraz przez uwzględnienie znaku $(-1)^{i+j}$.

### Macierz odwrotna 2x2

Analogicznie, dla macierzy 2x2, można obliczyć macierz odwrotną przy użyciu macierzy dopełnień algebraicznych:

$$
\mathbf{L}^{-1} =
\begin{bmatrix}
a & b \\
c & d
\end{bmatrix}^{-1}
= \frac{1}{\det(\mathbf{L})}
\begin{bmatrix}
D & C \\
B & A
\end{bmatrix}^T
= \frac{1}{\det(\mathbf{L})}
\begin{bmatrix}
D & B \\
C & A
\end{bmatrix}
$$

gdzie $\det(\mathbf{L}) = ad - bc$

Jeśli wyznacznik jest różny od zera to macierz jest odwracalna, a elementy macierzy pośredniej są określone jako:

$$
\begin{aligned}
A &= \det \begin{bmatrix} d \end{bmatrix} = d, & B &= -\det \begin{bmatrix} b \end{bmatrix} = -b, & C &= -\det \begin{bmatrix} c \end{bmatrix} = -c, &
D &= \det \begin{bmatrix} a \end{bmatrix} = a.
\end{aligned}
$$

Elementy te powstają przez obliczenie wyznaczników macierzy 1x1, które powstają przez usunięcie odpowiednich wierszy i kolumn z oryginalnej macierzy 2x2 oraz przez zastosowanie odpowiednich znaków $(-1)^{i+j}$.

---
## Zadania dla  studentów

1. Wylicz ze wzoru macierz odwrotną do macierzy

$$
\begin{bmatrix}
2 & 0 & 1 \\
0 & 1 & 0 \\
1 & 2 & 0
\end{bmatrix}
$$

i sprawdź, czy wynik jest poprawny.

2. Posługując się komendą `det()` ustal rząd macierzy poniżej
$$
A =
\begin{bmatrix}
4 & -3 & 7 \\
-1 & 6 & 3 \\
2 & 9 & 1
\end{bmatrix}
$$

In [ ]:
A = sp.Matrix([[2,0,1],[0,1,0],[1,2,0]]) #tworzymy macierz
A

Matrix([
[2, 0, 1],
[0, 1, 0],
[1, 2, 0]])

In [ ]:
inverse_A = A.inv() #odwracamy macierz
inverse_A

Matrix([
[0, -2,  1],
[0,  1,  0],
[1,  4, -2]])

In [ ]:
inverse_A*A #sprawdzamy czy mnożenie macierzy przez jej odwrotność daje macierz jednostkową

Matrix([
[1, 0, 0],
[0, 1, 0],
[0, 0, 1]])

In [ ]:
a,b,c,d,e,f,g,h,i = sp.symbols('a b c d e f g h i') #definiujemy zmienne a-i
A=sp.Matrix([[2, 0, 1], [0, 1, 0], [1, 2, 0]]) #tworzymy macierz A
inverse_A = k.inv()
inverse_A

Matrix([
[0, -2,  1],
[0,  1,  0],
[1,  4, -2]])

In [ ]:
A #wyświetlamy macierz A

Matrix([
[2, 0, 1],
[0, 1, 0],
[1, 2, 0]])

In [ ]:
A.transpose() #transponujemy macierz A (Transpozycja macierzy to operacja, która zamienia wiersze macierzy na kolumny i kolumny na wiersze)

Matrix([
[2, 0, 1],
[0, 1, 2],
[1, 0, 0]])

In [ ]:
A = sp.Matrix([[2, 0, 1], [0, 1, 0], [1, 2, 0]])
A

Matrix([
[2, 0, 1],
[0, 1, 0],
[1, 2, 0]])

In [ ]:
M11 = A.minor_submatrix(1, 1) #obliczamy minor M11 (Podwyznacznik lub minor macierzy to wyznacznik uzyskany z oryginalnej macierzy przez usunięcie jednego wiersza i jednej kolumny)
M11

Matrix([
[2, 1],
[1, 0]])

In [ ]:
M22 = A.minor_submatrix(2, 2) #obliczamy minor M22
M22

Matrix([
[2, 0],
[0, 1]])

In [ ]:
A = sp.Matrix([[4,-3,7],[-1,6,3],[2,9,1]]) #tworzymy macierz A
A

Matrix([
[ 4, -3, 7],
[-1,  6, 3],
[ 2,  9, 1]])

In [ ]:
det_A = A.det() #obliczamy wyznacznik macierzy A
det_A

-252

Dla macierzy 𝐴 obliczony wyznacznik wyniósł −252, co oznacza, że macierz jest pełnego rzędu 3